PurchasesCoreSwift/Networking/Backend.swift

```diff
@@ -0,0 +1,650 @@
+                return
+            }
+
+            guard let offers = response?["offers"] as? [[String: AnyObject]], offers.count > 0 else {
+                completion(nil, nil, nil, nil, ErrorUtils.unexpectedBackendResponseError())
+                return
+            }
+
+            let offer = offers[0]
+            if let signatureError = offer["signature_error"] as? [String: AnyObject] {
+                let code = self.maybeNumberFromError(code: signatureError["code"])
+                let backendMessage = signatureError["message"] as? String
+                let error = ErrorUtils.backendError(withBackendCode: code, backendMessage: backendMessage)
+                completion(nil, nil, nil, nil, error)
+
+            } else if let signatureData = offer["signature_data"] as? [String: AnyObject] {
+                let signature = signatureData["signature"] as? String
+                let keyIdentifier = offer["key_id"] as? String
+                let nonceString = signatureData["nonce"] as? String
+                let nonce: UUID?
+                if let nonceString = nonceString {
+                    nonce = UUID(uuidString: nonceString)
+                } else {
+                    nonce = nil
+                }
+
+                let timestamp = signatureData["timestamp"] as? NSNumber
+
+                completion(signature, keyIdentifier, nonce, timestamp, nil)
+                return
+            } else {
+                completion(nil, nil, nil, nil, ErrorUtils.unexpectedBackendResponseError())
+                return
+            }
+        }
+    }
+
+    @objc(postAttributionData:network:appUserID:completion:)
+    public func post(attributionData: [String: AnyObject],
+                     network: AttributionNetwork,
+                     appUserID: String,
+                     completion: ((Error?) -> Void)?) {
+        guard let appUserID = try? escapedAppUserID(appUserID: appUserID) else {
+            completion?(ErrorUtils.missingAppUserIDError())
+            return
+        }
+
+        let path = "/subscribers/\(appUserID)/attribution"
+        let body: [String: Any] = ["network": NSNumber(value: network.rawValue), "data": attributionData]
+        httpClient.performPOSTRequest(serially: true,
+                                      path: path,
+                                      requestBody: body,
+                                      headers: authHeaders) { statusCode, response, error in
+            self.handle(response: response, statusCode: statusCode, error: error, completion: completion)
+        }
+    }
+
+    @objc(postSubscriberAttributes:appUserID:completion:)
+    public func post(subscriberAttributes: SubscriberAttributeDict,
+                     appUserID: String,
+                     completion: ((Error?) -> Void)?) {
+        guard subscriberAttributes.count > 0 else {
+            Logger.warn(Strings.attribution.empty_subscriber_attributes)
+            return
```

Comment:
we should probably still call completion here. I think this was a bug in the previous implementation

Reply:
Yeah, the thing I was thinking about this piece was, is this an error condition? It doesn't seem like a hard error, but more like "hey, you might have expected attributes but there are none". That being said, calling the completion with a nil error also seems wrong.

Reply:
I decided to create another ErrorCode case